lib/vector-core/src/config/mod.rs

```diff
@@ -100,42 +100,132 @@ impl Input {
 }
 
 #[derive(Debug, Clone, PartialEq)]
-pub struct Output {
+pub struct SourceOutput {
     pub port: Option<String>,
     pub ty: DataType,
 
     // NOTE: schema definitions are only implemented/supported for log-type events. There is no
     // inherent blocker to support other types as well, but it'll require additional work to add
     // the relevant schemas, and store them separately in this type.
+    pub schema_definition: Option<schema::Definition>,
+}
+
+impl SourceOutput {
+    /// Create a `SourceOutput` of the given data type that contains a single output `Definition`.
+    /// Designed for use in log sources.
     ///
-    /// The `None` variant of a schema definition has two distinct meanings for a source component
-    /// versus a transform component:
-    ///
-    /// For *sources*, a `None` schema is identical to a `Some(Definition::source_default())`.
+    /// There is nothing intrinsically that forces this function to be used just for logs. `ty` could
+    /// be `DataType::Metrics` and this function will not complain. The only thing is this function
+    /// forces a `schema_definition` to be specified, which is required for logs.
+    #[must_use]
+    pub fn new_logs(ty: DataType, schema_definition: schema::Definition) -> Self {
+        Self {
+            port: None,
+            ty,
+            schema_definition: Some(schema_definition),
+        }
+    }
+
+    /// Create a `SourceOutput` of the given data type that contains no output `Definition`s.
+    /// Designed for use in metrics sources.
     ///
-    /// For a *transform*, a schema [`schema::Definition`] is required if `Datatype` is Log.
-    pub log_schema_definition: Option<schema::Definition>,
-}
+    /// There is nothing intrinsically that forces this function to be used just for metrics. `ty` could
```

Comment:
Concerning these `new_*` functions, should there be some programatic enforcement of type? Not sure if this is a valid concern, but it seems like someone could accidentally call `new_metrics()` with a log event. Would that cause problems?

Reply:
I could put some asserts in.

Reply:
I could put some asserts in.